pkg/kv/kvserver/store_send.go

```diff
@@ -157,104 +162,145 @@ func (s *Store) Send(
-		// uninitialized replica to be created is most likely the
-		// leader.
-		return nil, roachpb.NewError(&roachpb.NotLeaseHolderError{
-			RangeID:     ba.RangeID,
-			LeaseHolder: repl.creatingReplica,
-			// The replica doesn't have a range descriptor yet, so we have to build
-			// a ReplicaDescriptor manually.
-			Replica: roachpb.ReplicaDescriptor{
-				NodeID:    repl.store.nodeDesc.NodeID,
-				StoreID:   repl.store.StoreID(),
-				ReplicaID: replicaID,
-			},
-		})
-	}
-
-	br, pErr = repl.Send(ctx, ba)
-	if pErr == nil {
-		return br, nil
-	}
-
-	// Augment error if necessary and return.
-	switch t := pErr.GetDetail().(type) {
-	case *roachpb.RangeKeyMismatchError:
-		// TODO(andrei): It seems silly that, if the client specified a RangeID that
-		// doesn't match the keys it wanted to access, but this node can serve those
-		// keys anyway, we still return a RangeKeyMismatchError to the client
-		// instead of serving the request. Particularly since we have the mechanism
-		// to communicate correct range information to the client when returning a
-		// successful response (i.e. br.RangeInfos).
-
-		// On a RangeKeyMismatchError where the batch didn't even overlap
-		// the start of the mismatched Range, try to suggest a more suitable
-		// Range from this Store.
-		rSpan, err := keys.Range(ba.Requests)
-		if err != nil {
-			return nil, roachpb.NewError(err)
-		}
-
-		// The kvclient thought that a particular range id covers rSpans. It was
-		// wrong; the respective range doesn't cover all of rSpan, or perhaps it
-		// doesn't even overlap it. Clearly the client has a stale range cache.
-		// We'll return info on the range that the request ended up being routed to
-		// and, to the extent that we have the info, the ranges containing the keys
-		// that the client requested, and all the ranges in between.
-		ri, err := t.MismatchedRange()
+	// Tracks suggested ranges to return to the caller. Suggested ranges are aggregated from
+	// two sources.
+	// (1): On a RangeKeyMismatchError that is retriable.
+	// (2): On a successful batch request, where suggested ranges are returned
+	//      by the replica to update the client with. This is appended before returning.
+	var candidateRanges []roachpb.RangeInfo
+
+	// Run a retry loop on retriable RangeKeyMismatchErrors, where the requested RangeID is
+	// no longer correct, however we have the Range locally and can redirect the request.
+	// TODO(kvoli): The number of retries is limited to at most two repeated mismatch ranges. Recurrent
+	// RangeKeyMismatchErrors shouldn't be possible unless repeated splitting upon the range containing
+	// the request key span occurs. This seems unlikely, however as this behavior is degenerate retries
+	// is limited to at most 3 before returning an error to the client. Establish an appropriate bounds.
+	for retries, isRetriableRangeKeyMismatch := 0, true; retries < maxRKeyMismatchRetries && isRetriableRangeKeyMismatch; retries++ {
```

Comment:
Checkout pkg/util/retry -- I'm curious if we can use that library here instead, maybe getting rid of maxRKeyMismatchRetries. Seeing as how this retry block is useful only in the presence of recurrent splits, which if it were to occur would induce retries at the client anything (but with additional RTTs), it'd be fine IMO to retry until the surrounding ctx gets cancelled.